src/foam/nanos/crunch/ServerCrunchService.java

```diff
@@ -205,6 +207,51 @@ public UserCapabilityJunction getJunction(X x, String capabilityId) {
     return this.getJunctionForSubject(x, capabilityId, subject);
   }
 
+  public boolean hasPreconditionsMet(
+    X sessionX, String capabilityId
+  ) {
+    // Return false if capability does not exist or is not available
+    var capabilityDAO = ((DAO) sessionX.get("capabilityDAO")).inX(sessionX);
+    if ( capabilityDAO.find(capabilityId) == null ) return false;
+
+    var preconditions = Arrays.stream(((CapabilityCapabilityJunction[]) ((ArraySink) ((DAO) sessionX.get("prerequisiteCapabilityJunctionDAO"))
+      .where(AND(
+        EQ(CapabilityCapabilityJunction.SOURCE_ID, capabilityId),
+        EQ(CapabilityCapabilityJunction.PRECONDITION, true)
+      ))
+      .select(new ArraySink())).getArray()
+      .toArray(new CapabilityCapabilityJunction[0])))
+      .map(CapabilityCapabilityJunction::getTargetId).toArray(String[]::new);
+    
+    for ( String preconditionId : preconditions ) {
+      // Return false if capability does not exist or is not available
+      if ( capabilityDAO.find(preconditionId) == null ) return false;
+      var ucj = getJunction(sessionX, preconditionId);
+      if ( ucj.getStatus() != CapabilityJunctionStatus.GRANTED ) return false;
+    }
+
+    return true;
+  }
+
+  public ArraySink getEntryCapabilities(X x) {
```

Comment:
please add documentation to the purpose of this function and hasPreconditionsMet